Newton's Method
---------------------
Solve the following system for the Newton's method:

\begin{aligned}
f_1(x,y,z) &= x^2 + y + 2z^2 - 10 = 0 \\
f_2(x,y,z) &= 5x - 6y + z = 0 \\
f_3(x,y,z) &= z - x^2 - y^2 = 0
\end{aligned}


- Graph the functions 
- Get the roots with a tolerance that you define
- Use displacement successive and concurrent
- Show your results 

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme()

In [4]:
functions = [lambda x, y, z: z - x ** 2 - y ** 2,
     lambda x, y, z: 5 * x - 6 * y + z,
     lambda x,y,z: x**2 + y + 2*z**2 -10,]
   
f_derivative = [lambda x,y,z: -2*x,
               lambda x,y,z: -6,
               lambda x,y,z: 4*z]

Apply successive displacements.

In [6]:
x0 = [-1, -0.5, 2]
sol = x0.copy()
soluciones = [sol.copy()]
error_abs = [1]
while error_abs[-1] > 0.001:
  for i in range(len(x0)):
    sol[i] = sol[i] - functions[i](*sol) / f_derivative[i](*sol)
  soluciones.append(sol.copy())
  error_abs.append(np.max(np.abs(
      np.array(soluciones[-1]) - np.array(soluciones[-2])
  )))

error_abs[0] = np.nan
arreglo_soluciones = np.array(soluciones)

tabla = pd.DataFrame({
    'x aprox': arreglo_soluciones[:,0],
    'y aprox': arreglo_soluciones[:,1],
    'z aprox': arreglo_soluciones[:,2],
    'Error absoluto': error_abs,
})
tabla

x aprox   y aprox   z aprox  Error absoluto
0  -1.000000 -0.500000  2.000000             NaN
1  -1.375000 -0.812500  2.115234        0.375000
2  -1.216619 -0.661310  2.142739        0.158381
3  -1.309189 -0.733868  2.123748        0.092570
4  -1.260002 -0.696044  2.134087        0.049187
5  -1.284607 -0.714825  2.128927        0.024605
6  -1.272050 -0.705220  2.131563        0.012557
7  -1.278385 -0.710060  2.130234        0.006335
8  -1.275170 -0.707603  2.130909        0.003215
9  -1.276797 -0.708846  2.130568        0.001626
10 -1.275973 -0.708216  2.130740        0.000824

The method converges to the initial values we gave it, the first point $(-1.27625, -0.708428, 2.130686)$ is found in four iterations and the second point is found at $(0.906459, 1.090540, 2.010944)$.

Apply concurrent displacement .

In [8]:
x0 = [-1,-0.5,2]
sol = x0.copy()
solP=sol.copy()
soluciones = [sol.copy()]
error_abs = [1]
while error_abs[-1] > 0.001:
  for i in range(len(x0)):
    solP[i] = sol[i] - functions[i](*sol) / f_derivative[i](*sol)
  soluciones.append(solP.copy())
  error_abs.append(np.max(np.abs(
      np.array(soluciones[-1]) - np.array(soluciones[-2])
  )))
  sol = solP 

error_abs[0] = np.nan
arreglo_soluciones = np.array(soluciones)

tabla = pd.DataFrame({
    'x aprox': arreglo_soluciones[:,0],
    'y aprox': arreglo_soluciones[:,1],
    'z aprox': arreglo_soluciones[:,2],
    'Error absoluto': error_abs,
})
tabla

x aprox   y aprox   z aprox  Error absoluto
0  -1.000000 -0.500000  2.000000             NaN
1  -1.375000 -0.500000  2.187500        0.375000
2  -1.392045 -0.795455  2.106054        0.295455
3  -1.225210 -0.670000  2.141420        0.166835
4  -1.303311 -0.729189  2.124983        0.078101
5  -1.262893 -0.698247  2.133482        0.040418
6  -1.283098 -0.713668  2.129244        0.020205
7  -1.272803 -0.705795  2.131405        0.010295
8  -1.278000 -0.709766  2.130315        0.005197
9  -1.275364 -0.707751  2.130868        0.002636
10 -1.276698 -0.708771  2.130588        0.001334
11 -1.276022 -0.708254  2.130730        0.000676

We can see that we get the root (-1.2760, -0.7082, 2.1307) in the 11 iteration with an error of 0.000676

Now, we need to find the positive root, however, is necessary to change the order of functions, otherwise, the method gives us the root that we have found yet. 

In [9]:
# definimos nuestras funciones

f2 = [lambda x, y, z: 5 * x - 6 * y + z,
      lambda x, y, z: z - x ** 2 - y ** 2,
      lambda x, y, z: x**2 + y + 2*z**2 - 10,]

# declaramos las derivadas
f2_derivadas = [lambda x, y, z: 5,
                lambda x, y, z: -2*y,
               lambda x, y,z: 4*z]

Apply concurrent displacement.

In [11]:
x0 = [1, 1, 2]
sol = x0.copy()
solP = sol.copy()
soluciones = [sol.copy()]
error_abs = [1]
while error_abs[-1] > 0.001:
  for i in range(len(x0)):
    solP[i] = sol[i] - f2[i](*sol) / f2_derivadas[i](*sol)
  soluciones.append(solP.copy())
  error_abs.append(np.max(np.abs(
      np.array(soluciones[-1]) - np.array(soluciones[-2])
  )))
  sol = solP

error_abs[0] = np.nan
arreglo_soluciones = np.array(soluciones)

tabla = pd.DataFrame({
    'x aprox': arreglo_soluciones[:, 0],
    'y aprox': arreglo_soluciones[:, 1],
    'z aprox': arreglo_soluciones[:, 2],
    'Error absoluto': error_abs,
})
tabla.tail()

x aprox   y aprox   z aprox  Error absoluto
46  0.905743  1.091087  2.011037        0.001520
47  0.907097  1.090052  2.010861        0.001354
48  0.905891  1.090974  2.011018        0.001206
49  0.906965  1.090153  2.010878        0.001074
50  0.906008  1.090884  2.011003        0.000957

We can see that with concurrent displacement the solution is (0.9060, 1.0908, 2.0110) in 50 iterations with an error of 0.000957.

Apply successive displacements.

In [12]:
x0 = [1, 1, 2]
sol = x0.copy()
soluciones = [sol.copy()]
error_abs = [1]
while error_abs[-1] > 0.001:
  for i in range(len(x0)):
    sol[i] = sol[i] - f2[i](*sol) / f2_derivadas[i](*sol)
  soluciones.append(sol.copy())
  error_abs.append(np.max(np.abs(
      np.array(soluciones[-1]) - np.array(soluciones[-2])
  )))

error_abs[0] = np.nan
arreglo_soluciones = np.array(soluciones)

tabla = pd.DataFrame({
    'x aprox': arreglo_soluciones[:,0],
    'y aprox': arreglo_soluciones[:,1],
    'z aprox': arreglo_soluciones[:,2],
    'Error absoluto': error_abs,
})
tabla.tail()

x aprox   y aprox   z aprox  Error absoluto
45  0.905743  1.091087  2.011037        0.001520
46  0.907097  1.090052  2.010861        0.001354
47  0.905891  1.090974  2.011018        0.001206
48  0.906965  1.090153  2.010878        0.001074
49  0.906008  1.090884  2.011003        0.000957

The positive root was found after 49 iterations, returning the point (0.906008,1.090884,2.011003) with an error of 0.00095.